cmake/modules/RootMacros.cmake

```diff
@@ -283,115 +283,226 @@ function(ROOT_GENERATE_DICTIONARY dictionary)
+       list(APPEND incdirs ${CMAKE_SOURCE_DIR}/io/io/inc)
+    endif()
+
+    set(headerdirs)
 
-  if(TARGET ${ARG_MODULE})
     get_target_property(target_incdirs ${ARG_MODULE} INCLUDE_DIRECTORIES)
-    foreach(dir ${target_incdirs})
-      string(REGEX REPLACE "^[$]<BUILD_INTERFACE:(.+)>" "\\1" dir ${dir})
-      if(NOT ${dir} MATCHES "^[$]")
-        list(APPEND incdirs ${dir})
-      endif()
+    if(target_incdirs)
+       foreach(dir ${target_incdirs})
+          string(REGEX REPLACE "^[$]<BUILD_INTERFACE:(.+)>" "\\1" dir ${dir})
+          if(NOT ${dir} MATCHES "^[$]")
+             list(APPEND incdirs ${dir})
+          endif()
+          if(${dir} MATCHES "^${CMAKE_SOURCE_DIR}")
+             list(APPEND headerdirs ${dir})
+          endif()
+       endforeach()
+    endif()
+
+
+    # Comments from Sergey
+    # FIXME: we have to exclude all source directories from includes while they duplicated in ${CMAKE_BINARY_DIR}/include
+    # rootcling is not able to handle such situation, therefore ${CMAKE_BINARY_DIR}/include is used as source for ROOT headers
+    # Also ${CMAKE_BINARY_DIR}/ginclude is removed
+    # add slash at the end to prevent filtering include paths from roottest which may have same location as root
+    list(FILTER incdirs EXCLUDE REGEX "^${CMAKE_SOURCE_DIR}\/")
+    list(FILTER incdirs EXCLUDE REGEX "^${CMAKE_BINARY_DIR}/ginclude")
+    list(FILTER incdirs EXCLUDE REGEX "^${CMAKE_BINARY_DIR}/externals")
+    list(INSERT incdirs 0 ${CMAKE_BINARY_DIR}/include)
+    # end workaround
+
+    if(ARG_MODULE STREQUAL Core)
+       list(APPEND incdirs ${CMAKE_SOURCE_DIR}/core) # This is needed because LinkDef.h includes other LinkDef starting from ${CMAKE_SOURCE_DIR}/core
+    endif()
+
+    if(incdirs)
+       list(REMOVE_DUPLICATES incdirs)
+    endif()
+
+    set(includedirs)
+    foreach(dir ${incdirs})
+       list(APPEND includedirs -I${dir})
     endforeach()
-  endif()
 
-  #---Get the list of header files-------------------------
-  # CMake needs dependencies from ${CMAKE_CURRENT_SOURCE_DIR} while rootcling wants
-  # header files "as included" (and thus as passed as argument to this CMake function).
-  set(headerfiles)
-  set(_list_of_header_dependencies)
-  foreach(fp ${ARG_UNPARSED_ARGUMENTS})
-    if(${fp} MATCHES "[*?]") # Is this header a globbing expression?
-      file(GLOB files inc/${fp} ${fp}) # Elements of ${fp} have the complete path.
-      foreach(f ${files})
-        if(NOT f MATCHES LinkDef) # skip LinkDefs from globbing result
-          set(add_inc_as_include On)
-          string(REGEX REPLACE "^${CMAKE_CURRENT_SOURCE_DIR}/inc/" "" f_no_inc ${f})
-          list(APPEND headerfiles ${f_no_inc})
-          list(APPEND _list_of_header_dependencies ${f})
+    set(pureincdirs ${incdirs})
+
+    set(headerfiles)
+    set(_list_of_header_dependencies)
+    foreach(fp ${ARG_UNPARSED_ARGUMENTS})
+       if(IS_ABSOLUTE ${fp})
+          set(headerFile ${fp})
+       else()
+          find_file(headerFile ${fp}
```

Comment:
Do you use `find_file` to get the full path to a given header file?

Reply:
Yes, exactly as before